### PR TITLE
Fix compatibility with Matplotlib 3.5

### LINKED ISSUE #7272: Terra incompatible with `matplotlib==3.5.0`
### Environment

- **Qiskit Terra version**:  0.19.0.dev @ 7e118e8
- **Python version**: 3.7+
- **Operating system**: any

### What is happening?

CI is currently failing on all docs and Linux 3.7 builds (only this one because the image tests are in it).

The docs failure is:
```python
Extension error:
Cell raised uncaught exception:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/work/1/s/.tox/docs/lib/python3.7/site-packages/IPython/core/formatters.py in __call__(self, obj)
    339                 pass
    340             else:
--> 341                 return printer(obj)
    342             # Finally look for special method names
    343             method = get_real_method(obj, self.print_method)

~/work/1/s/.tox/docs/lib/python3.7/site-packages/IPython/core/pylabtools.py in <lambda>(fig)
    437 
    438             collections_and_patches = (

AttributeError: 'Arrow3D' object has no attribute 'do_3d_projection'
```

and the Linux py3.7 failure is:
```python
ERROR: test_plot_state_qsphere (mpl.graph.test_graph_matplotlib_drawer.TestGraphMatplotlibDrawer)
test for plot_state_qsphere
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py", line 139, in test_plot_state_qsphere
    self.graph_state_drawer(state=state, output="qsphere", filename="qsphere.png")
  File "/home/vsts/work/1/s/test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py", line 84, in wrapper
    results = func(*args, **kwargs)
  File "/home/vsts/work/1/s/qiskit/visualization/state_visualization.py", line 1433, in state_drawer
    return draw_func(state, **drawer_args)
  File "/home/vsts/work/1/s/qiskit/utils/deprecation.py", line 27, in wrapper
    return func(*args, **kwargs)
  File "/home/vsts/work/1/s/qiskit/visualization/state_visualization.py", line 961, in plot_state_qsphere
    return fig.savefig(filename)
  File "/home/vsts/work/1/s/image_tests/lib/python3.7/site-packages/matplotlib/figure.py", line 3012, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/home/vsts/work/1/s/image_tests/lib/python3.7/site-packages/matplotlib/_api/__init__.py", line 299, in warn_external
    warnings.warn(message, category, stacklevel)
matplotlib._api.deprecation.MatplotlibDeprecationWarning: The 'renderer' parameter of do_3d_projection() was deprecated in Matplotlib 3.4 and will be removed two minor releases later.
```

### How can we reproduce the issue?

Check a recent CI run.

### What should happen?

No error.

### Any suggestions?

Immediately, pin `matplotlib<3.5` to get CI rolling, then work on fixing compatibility.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

The problem was `qiskit.visualization.bloch.Arrow3D`, which subclassed
`matplotlib.patches.FancyArrowPatch` (a 2D patch), but not the 3D
machinery.  It is now made an unholy multiple-inherited abomination of
both the 2D patch and `mpl_toolkits.mplot3d.art3d.Patch3D`; the latter
is a relatively thin wrapper (in terms of attributes) around the 2D
patch, so this is not too terrible.

Matplotlib 3.5 calls the `Patch3D.do_3d_projection` method using a
deprecated parameter, triggering two warnings, unless the artist's
module appears to have come from `mpl_toolkits.mplot3d.art3d`, even if
the new calling convention is respected.  The warning is only triggered
when the figure is drawn, which may well be outside of our control, so
we cannot suppress the warnings.  Instead, we just lie about the module
the arrow patch was defined in, to trick it into not warning, because we
use the new calling convention.  This is supported at least as far back
as Matplotlib 3.3, which is the current minimum supported version.  The
nasty hack should be removable once Matplotlib 3.6 is the minimum
version because the deprecation period will expire.

### Details and comments

This isn't neat or pretty, but it's an attempt to keep the behaviour exactly as it has been ever since this file came from qutip, since I don't understand `mplot3d` very well.  Fix #7272.

### CODE REVIEW

File: qiskit/visualization/bloch.py [hunk 1]
```diff
@@ -54,19 +54,36 @@
...
-        FancyArrowPatch.__init__(self, (0, 0), (0, 0), *args, **kwargs)
-        self._verts3d = xs, ys, zs
+    # Nasty hack around a poorly implemented deprecation warning in Matplotlib 3.5 that issues two
+    # deprecation warnings if an artist's module does not claim to be part of the below module.
+    # This revolves around the method `Patch3D.do_3d_projection(self, renderer=None)`.  The
+    # `renderer` argument has been deprecated since Matplotlib 3.4, but in 3.5 some internal calls
+    # during `Axes3D` display started calling the method.  If an artist does not have this module,
+    # then it issues a deprecation warning, and calls it by passing the `renderer` parameter as
+    # well, which consequently triggers another deprecation warning.  We should be able to remove
+    # this once 3.6 is the minimum supported version, because the deprecation period ends then.
```
Comment (mtreinish) [inline qiskit/visualization/bloch.py:72]: Fwiw, it'll be a while before we can make our matplotlib requirement `>=3.6`.

File: qiskit/visualization/bloch.py [hunk 2]
```diff
@@ -54,19 +54,36 @@
...
+    __module__ = "mpl_toolkits.mplot3d.art3d"
+
+    def __init__(self, xs, ys, zs, zdir="z", **kwargs):
+        # The Patch3D.__init__() method just calls its own super() method and then
+        # self.set_3d_properties, but it its __init__ signature is actually pretty incompatible with
```
Comment (mtreinish) [inline qiskit/visualization/bloch.py:77]:
```suggestion
        # self.set_3d_properties, but its __init__ signature is actually pretty incompatible with
```

**Review by mtreinish — DISMISSED** [review]
LGTM, one inline comment typo to fix. I really do think this is worth opening an issue in mpl about (if you haven't already) since this seems like a pretty broken migration/deprecation plan for people. Especially since the `Arrow3D` code is basically identical in every search result for "3d arrow matplotlib".

File: qiskit/visualization/bloch.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/visualization/bloch.py:72]: Yeah I know - here it's just that I don't like putting in gross hacks without saying the condition that they can be removed again!

**Comment by jakelishman** [comment]
Yeah, I'll make the issue on MPL.  I didn't before because I didn't know enough about what was going on, but I've read the `mplot3d` source now and worked out what we were trying to do, so I'm happier that there's no way to faithfully avoid the warning being triggered, and since we're a library and the draw call happens outside our control, we can't apply warning filters without affecting the user's global state.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**